File: src/kohakulayout/state/wiring.py
"""The wire side of the world: route a net, keep or take up its wire, and the units that go with it.

A mixin of ``World``; it reads the world's kernel, wires, units and router and writes
through the world's undo log like every other mutation.
"""

from collections.abc import Iterable
from typing import Any

from kohakulayout.errors import StateError
from kohakulayout.ir import Refusal, Wire
from kohakulayout.ir.geometry import XY
from kohakulayout.state.crossing import side_bit
from kohakulayout.state.kernel import holder_kind


class WiringMixin:
    """Routing and wires: the world keeps the record, the router decides the paths."""

    netlist: Any
    kernel: Any
    router: Any
    wires: dict[str, Wire]
    units: dict[str, Any]

    def route(self, net_id: str, grow: bool = False) -> Refusal | None:
        """Route a net; with ``grow`` a routed net keeps its tree and reaches the pins it does not touch yet."""
        if net_id not in self.netlist.nets:
            raise StateError(f"no net {net_id!r}")
        if net_id in self.wires and not grow:
            return None
        pre_digest = self.digest() if self.checker is not None else ""
        if self.router is None:
            result: Refusal | None = Refusal(
                stage="route", subject=f"net:{net_id}", detail="no router is installed"
            )
        else:
            mark = self.mark()
            result = self.router.route(self, net_id)
            if result is not None:
                self.rollback_to(mark)
        if self.checker is not None:
            self.checker.on_route(self, net_id, result, pre_digest)
        return result

    def forget_routes(self) -> None:
        """Tell the router its negotiation history no longer describes this state."""
        self._tables = None
        forget = getattr(self.router, "forget", None)
        if forget is not None:
            forget()

    def runs_of(self, wire: Wire) -> dict[str, list[tuple[XY, int]]]:
        """Per layer, the sides the wire continues to on each cell, a used port behind included."""
        out: dict[str, dict[XY, int]] = {}
        for segment in wire.segments:
            table = out.setdefault(segment.layer, {})
            cells = segment.cells
            for i, xy in enumerate(cells):
                mask = table.get(xy, 0)
                if i > 0:
                    mask |= side_bit(xy, cells[i - 1])
                if i + 1 < len(cells):
                    mask |= side_bit(xy, cells[i + 1])
                table[xy] = mask
        net = self.netlist.nets[wire.net]
        table = out.get(self.carrier_layer(net.carrier))
        if table:
            for ref in net.pins():
                found = (
                    self.choice(ref.cell, ref.pin)
                    if ref.cell in self.placements
                    else None
                )
                if found is not None and found[1] in table:
                    table[found[1]] |= side_bit(found[1], found[2])
        return {layer: sorted(table.items()) for layer, table in out.items()}

    def set_wire(self, wire: Wire) -> None:
        """Record a routed wire, hold its cells and write its runs."""
        for segment in wire.segments:
            self._occupy((segment.layer,), tuple(segment.cells), f"wire:{wire.net}")
        self.wires[wire.net] = wire
        self.retable(wire.net)
        self._record(lambda: (self.wires.pop(wire.net, None), self.retable(wire.net)))
        self._write_runs(f"wire:{wire.net}", self.runs_of(wire), {})

    def trim(self, net_id: str, cells: Iterable[XY]) -> bool:
        """Take up the segments ``cells`` cut and every segment hanging from them.

        A segment hangs from a taken one when an end of it lies on that segment's own cells; the
        rest stays a routed wire. False when nothing stays.
        """
        wire = self.wires.get(net_id)
        if wire is None:
            return False
        hit = frozenset(cells)
        segments = list(wire.segments)
        gone = [seg for seg in segments if hit.intersection(seg.cells)]
        if not gone:
            return False
        before = self.runs_of(wire)
        ports = {
            xy
            for key, port_id in wire.ports.items()
            if (xy := self._attach_of(key, port_id)) is not None
        }
        for ref in self.netlist.nets[net_id].pins():
            if ref.cell in self.placements:
                ports.update(
                    a
                    for _, a, _ in self.port_choices(ref.cell).get(ref.pin, ())
                    if self.netlist.pin(ref) is None
                    or len(self.netlist.pin(ref).ports) <= 1
                )

        def hangs_at(i: int, end: XY) -> bool:
            """Whether a segment's end lies on an earlier lane, or on any other off a port cell."""
            if any(end in other.cells for other in segments[:i]):
                return True
            return end not in ports and any(
                end in other.cells for j, other in enumerate(segments) if j != i
            )

        leaves = [hangs_at(i, seg.cells[0]) for i, seg in enumerate(segments)]
        joins = [hangs_at(i, seg.cells[-1]) for i, seg in enumerate(segments)]

        def own(i: int) -> set[XY]:
            """A segment's own cells: the end it leaves from or joins at belongs to the earlier lane there."""
            cells = set(segments[i].cells)
            if leaves[i]:
                cells.discard(segments[i].cells[0])
            if joins[i]:
                cells.discard(segments[i].cells[-1])
            return cells

        while True:
            lost = set().union(
                *(own(i) for i, seg in enumerate(segments) if seg in gone)
            )
            more = [
                seg
                for i, seg in enumerate(segments)
                if seg not in gone
                and (
                    (leaves[i] and seg.cells[0] in lost)
                    or (joins[i] and seg.cells[-1] in lost)
                )
            ]
            if not more:
                break
            gone += more
        lost = {c for seg in gone for c in seg.cells}
        kept = [seg for seg in segments if seg not in gone]
        if not kept:
            return False
        staying = {c for seg in kept for c in seg.cells}
        freed = lost - staying
        cut = (
            {seg.cells[0] for seg in gone} | {seg.cells[-1] for seg in gone}
        ) & staying
        for layer in {seg.layer for seg in gone}:
            layer_cells = tuple(
                sorted(
                    c
                    for seg in gone
                    if seg.layer == layer
                    for c in seg.cells
                    if c in freed
                )
            )
            if layer_cells:
                self._free((layer,), layer_cells, f"wire:{net_id}")
        units = []
        for unit_id in wire.units:
            unit = self.units.get(unit_id)
            if unit is None:
                continue
            at = (unit.x, unit.y)
            if at in freed or (
                at in cut
                and not any(at in (seg.cells[0], seg.cells[-1]) for seg in kept)
            ):
                self.remove_unit(unit_id)
            else:
                units.append(unit_id)
        ports = {
            key: port_id
            for key, port_id in wire.ports.items()
            if self._attach_of(key, port_id) not in freed
        }
        trimmed = wire.model_copy(
            update={"segments": tuple(kept), "units": tuple(units), "ports": ports}
        )
        self.wires[net_id] = trimmed
        self.retable(net_id)
        self._record(
            lambda: (self.wires.__setitem__(net_id, wire), self.retable(net_id))
        )
        self._write_runs(f"wire:{net_id}", self.runs_of(trimmed), before)
        for layer in {seg.layer for seg in gone}:
            for xy in sorted(freed):
                for holder in self.kernel.holders_at(layer, xy):
                    kind, ref = holder_kind(holder)
                    if kind == "unit":
                        self._drop_shared_unit(ref, net_id)
        return True

    def _attach_of(self, key: str, port_id: str) -> XY | None:
        """The attach cell of a pin's recorded port, from its ``cell.pin`` key."""
        cell_id, pin_id = key.split(".", 1)
        for found, xy, _ in self.port_choices(cell_id).get(pin_id, ()):
            if found == port_id:
                return xy
        return None

    def unroute(self, net_id: str) -> None:
        """Free the wire, its units, and every other net's unit that only existed because of it."""
        wire = self.wires.get(net_id)
        if wire is None:
            return
        self._write_runs(f"wire:{net_id}", {}, self.runs_of(wire))
        held = self.kernel.cells_of(f"wire:{net_id}")
        for layer, cells in held.items():
            self._free((layer,), tuple(sorted(cells)), f"wire:{net_id}")
        for unit_id in list(wire.units):
            self.remove_unit(unit_id)
        del self.wires[net_id]
        self.retable(net_id)
        self._record(
            lambda: (self.wires.__setitem__(net_id, wire), self.retable(net_id))
        )
        for layer, cells in held.items():
            for xy in sorted(cells):
                for holder in self.kernel.holders_at(layer, xy):
                    kind, ref = holder_kind(holder)
                    if kind == "unit":
                        self._drop_shared_unit(ref, net_id)

    def _drop_shared_unit(self, unit_id: str, gone: str) -> None:
        """A unit another net's wire lists, on a cell the ripped net held, was a crossing: it goes too."""
        unit = self.units.get(unit_id)
        if (
            unit is None
            or not unit.owner.startswith("net:")
            or unit.owner == f"net:{gone}"
        ):
            return
        other_id = unit.owner.removeprefix("net:")
        other = self.wires.get(other_id)
        if other is None or unit_id not in other.units:
            return
        self.remove_unit(unit_id)
        trimmed = other.model_copy(
            update={"units": tuple(u for u in other.units if u != unit_id)}
        )
        self.wires[other_id] = trimmed
        self._record(lambda: self.wires.__setitem__(other_id, other))

File: src/kohakulayout-rs/src/kernel/wiring.rs
//! The wire bookkeeping of the simulated world: a wire's runs, a routed wire recorded or taken up
//! with the units only it needed, a unit taken away, and the seed and standing lanes a grow reads.

use std::collections::BTreeMap;

use super::grid::XY;
use super::hash::{Map, Set};
use super::judge::{holders, kind_of};
use super::lanes::{attach_pins, whole_lanes, Pair};
use super::records::Wire;
use super::sim::{footprint_cells, Decline, Sim};
use super::spec::Seed;

fn side(a: XY, b: XY) -> u8 {
    match (b.0 - a.0, b.1 - a.1) {
        (0, -1) => 1,
        (1, 0) => 2,
        (0, 1) => 4,
        (-1, 0) => 8,
        _ => 0,
    }
}

impl<'a> Sim<'a> {
    /// Per layer, the sides the wire continues to on each cell, a used port behind included.
    pub fn runs_of(
        &self,
        net: &str,
        wire: &Wire,
    ) -> Result<BTreeMap<String, Map<XY, u8>>, Decline> {
        let mut out: BTreeMap<String, Map<XY, u8>> = BTreeMap::default();
        for (layer, cells) in &wire.segments {
            let table = out.entry(layer.clone()).or_default();
            for (i, xy) in cells.iter().enumerate() {
                let mut mask = table.get(xy).copied().unwrap_or(0);
                if i > 0 {
                    mask |= side(*xy, cells[i - 1]);
                }
                if i + 1 < cells.len() {
                    mask |= side(*xy, cells[i + 1]);
                }
                table.insert(*xy, mask);
            }
        }
        let n = self.rec.nets.get(net).ok_or("an unknown net")?;
        let layer = self.layer_of(&n.carrier)?;
        let mut bits: Vec<(XY, u8)> = Vec::new();
        if out.get(layer).is_some_and(|t| !t.is_empty()) {
            for id in n.sources.iter().chain(&n.sinks) {
                let pin = self.rec.pins.get(id).ok_or("an unknown pin")?;
                if !self.placed(&pin.cell) {
                    continue;
                }
                if let Some((_, attach, port_cell)) = self.choice(&pin.cell, &pin.pin) {
                    bits.push((attach, side(attach, port_cell)));
                }
            }
        }
        if let Some(table) = out.get_mut(layer) {
            for (attach, bit) in bits {
                if let Some(mask) = table.get_mut(&attach) {
                    *mask |= bit;
                }
            }
        }
        Ok(out)
    }

    pub fn write_runs(
        &mut self,
        holder: &str,
        runs: &BTreeMap<String, Map<XY, u8>>,
        before: &BTreeMap<String, Map<XY, u8>>,
    ) {
        let layers: std::collections::BTreeSet<&String> =
            runs.keys().chain(before.keys()).collect();
        let empty = Map::default();
        for layer in layers {
            let new = runs.get(layer).unwrap_or(&empty);
            let old = before.get(layer).unwrap_or(&empty);
            let cells: Set<XY> = new.keys().chain(old.keys()).cloned().collect();
            let entries: Vec<(XY, u8)> = cells
                .iter()
                .map(|xy| (*xy, new.get(xy).copied().unwrap_or(0)))
                .collect();
            let undo: Vec<(XY, u8)> = cells
                .iter()
                .map(|xy| (*xy, old.get(xy).copied().unwrap_or(0)))
                .collect();
            self.set_runs(layer, holder, &entries, undo);
        }
    }

    /// Record a routed wire: hold its cells, retable its pins, write its runs.
    pub fn set_wire(&mut self, net: &str, wire: Wire) -> Result<(), Decline> {
        let holder = format!("wire:{net}");
        for (layer, cells) in &wire.segments {
            self.occupy(layer, cells, &holder);
        }
        self.wires.insert(net.to_string(), wire.clone());
        self.retable(net);
        let runs = self.runs_of(net, &wire)?;
        self.write_runs(&holder, &runs, &BTreeMap::default());
        Ok(())
    }

    /// Free the net's wire, its units, and the other units that stood only because of it.
    pub fn unroute(&mut self, net: &str) -> Result<(), Decline> {
        let Some(wire) = self.wires.get(net).cloned() else {
            return Ok(());
        };
        let holder = format!("wire:{net}");
        let before = self.runs_of(net, &wire)?;
        self.write_runs(&holder, &BTreeMap::default(), &before);
        let held = self.grid.cells_of(&holder);
        for (layer, cells) in &held {
            self.free(layer, cells, &holder);
        }
        for unit in &wire.units {
            self.remove_unit(unit)?;
        }
        self.wires.remove(net);
        self.retable(net);
        let shared: Vec<String> = held
            .iter()
            .flat_map(|(l, cells)| self.units_on(l, cells))
            .collect();
        self.drop_shared_units(&shared, net)
    }

    /// The units standing on the cells of a layer, cell by cell.
    pub fn units_on(&self, layer: &str, cells: &[XY]) -> Vec<String> {
        cells
            .iter()
            .flat_map(|xy| holders(self.grid, layer, *xy))
            .filter_map(|h| match kind_of(h) {
                ("unit", reference) => Some(reference.to_string()),
                _ => None,
            })
            .collect()
    }

    /// Drop each of the units that stood only because of `net`'s wire.
    pub fn drop_shared_units(&mut self, units: &[String], net: &str) -> Result<(), Decline> {
        for unit in units {
            self.drop_shared_unit(unit, net)?;
        }
        Ok(())
    }

    pub fn drop_shared_unit(&mut self, unit: &str, gone: &str) -> Result<(), Decline> {
        let Some(found) = self.units.get(unit) else {
            return Ok(());
        };
        let Some(other) = found.owner.strip_prefix("net:") else {
            return Ok(());
        };
        if other == gone {
            return Ok(());
        }
        let other = other.to_string();
        let Some(wire) = self.wires.get(&other) else {
            return Ok(());
        };
        if !wire.units.iter().any(|u| u == unit) {
            return Ok(());
        }
        let mut trimmed = wire.clone();
        trimmed.units.retain(|u| u != unit);
        self.remove_unit(unit)?;
        self.wires.insert(other, trimmed);
        Ok(())
    }

    pub fn layers_for(&self, footprint: &str) -> Result<Vec<String>, Decline> {
        let fp = self
            .rec
            .footprints
            .get(footprint)
            .ok_or("an unknown footprint")?;
        let mut out = vec![fp.layer.clone()];
        out.extend(fp.occludes.iter().cloned());
        Ok(out)
    }

    pub fn remove_unit(&mut self, unit: &str) -> Result<(), Decline> {
        let Some(found) = self.units.get(unit).cloned() else {
            return Ok(());
        };
        let fp = self
            .rec
            .footprints
            .get(&found.footprint)
            .ok_or("an unknown footprint")?;
        let cells = footprint_cells(found.x, found.y, fp.width, fp.height, found.rot);
        let holder = format!("unit:{unit}");
        for layer in self.layers_for(&found.footprint)? {
            self.free(&layer, &cells, &holder);
        }
        self.units.remove(unit);
        Ok(())
    }

    /// The seed a routed net gives a grow: its segments, junctions, crossings and ports.
    pub fn seed_of(&self, net: &str) -> Result<Option<Seed>, Decline> {
        let Some(wire) = self.wires.get(net) else {
            return Ok(None);
        };
        let n = self.rec.nets.get(net).ok_or("an unknown net")?;
        let carrier = self
            .rec
            .carriers
            .get(&n.carrier)
            .ok_or("an unknown carrier")?;
        let mut kinds: Vec<(&str, &str)> = Vec::new();
        for (name, fp) in [("split", &carrier.split), ("merge", &carrier.merge)] {
            if let Some(fp) = fp {
                kinds.retain(|(f, _)| *f != fp.as_str());
                kinds.push((fp.as_str(), name));
            }
        }
        let mut junctions: Vec<(XY, String)> = Vec::new();
        for unit in &wire.units {
            let Some(u) = self.units.get(unit) else {
                continue;
            };
            if let Some((_, name)) = kinds.iter().find(|(f, _)| *f == u.footprint) {
                let xy = (u.x, u.y);
                match junctions.iter_mut().find(|(c, _)| *c == xy) {
                    Some(entry) => entry.1 = name.to_string(),
                    None => junctions.push((xy, name.to_string())),
                }
            }
        }
        let layer = self.layer_of(&n.carrier)?;
        let cells: std::collections::BTreeSet<XY> = wire
            .segments
            .iter()
            .flat_map(|(_, c)| c.iter().cloned())
            .collect();
        let mut crossings: Vec<(XY, String, bool)> = Vec::new();
        for xy in cells {
            for h in holders(self.grid, layer, xy) {
                let (kind, reference) = kind_of(h);
                if kind == "wire" && reference != net {
                    crossings.push((xy, reference.to_string(), false));
                    break;
                }
            }
        }
        let own = self.tables.pair(&n.carrier, &n.carrier);
        if own.mode == 2 && !own.unit.is_empty() {
            let seen: Set<XY> = crossings.iter().map(|c| c.0).collect();
            for unit in &wire.units {
                let Some(u) = self.units.get(unit) else {
                    continue;
                };
                if u.footprint != own.unit {
                    continue;
                }
                if !seen.contains(&(u.x, u.y)) {
                    crossings.push(((u.x, u.y), net.to_string(), false));
                }
            }
        }
        Ok(Some(Seed {
            segments: wire.segments.iter().map(|(_, c)| c.clone()).collect(),
            junctions,
            crossings,
            ports: wire.ports.clone(),
        }))
    }

    /// The lanes the net's standing wire carries whole.
    pub fn standing_lanes(&self, net: &str) -> Result<Set<Pair>, Decline> {
        let Some(wire) = self.wires.get(net) else {
            return Ok(Set::default());
        };
        if wire.segments.is_empty() {
            return Ok(Set::default());
        }
        let n = self.rec.nets.get(net).ok_or("an unknown net")?;
        let Ok(found) = self.terminals(net)? else {
            return Ok(Set::default());
        };
        let sources: Set<&str> = n.sources.iter().map(String::as_str).collect();
        let at = attach_pins(&found, &wire.ports);
        let segments: Vec<Vec<XY>> = wire.segments.iter().map(|(_, c)| c.clone()).collect();
        let (_, read) = whole_lanes(segments, &at, &sources);
        Ok(read.into_iter().collect())
    }
}
